text/0000-pointee.md

```diff
@@ -0,0 +1,160 @@
+- Feature Name: `pointee`
+- Start Date: 2020-09-09
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+Add two language traits `Pointee` and `DynSized` to allow for more fine-grained type sizing.
+
+# Motivation
+[motivation]: #motivation
+
+As of right now, the lack of anything more fine-grained than `Sized` means Rust can deal with only two cases: a `Sized` type whose pointers and references are always thin, or a `!Sized` type whose pointers and references are always wide.
+
+There is thus no way to encode extern types as in [RFC #1861], nor any custom dynamically sized type in the far future with [RFC #2594].
+
+This RFC focuses on the necessary steps to be able to specify and implement extern types and custom dynamically sized types in the future.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Two new language traits are added to the standard library under `core::marker`:
+
+```rust
+/// Types whose values can be the target of a pointer.
+///
+/// As all types can be used in a pointer or reference type, this trait is implemented
+/// automatically by the compiler for all types.
+trait Pointee {
+    /// The pointer metadata associated with values of this type.
+    ///
+    /// It is the unit type `()` for `Sized` and `extern` types, `usize` for slice types,
+    /// and an unspecified opaque type for `dyn` types.
+    ///
+    /// Pointer types `*const T` and `*mut T` implement `Copy`, `Eq`, `Ord` and `Unpin`,
+    /// so the pointer metadata must implement too. Furthermore, reference types `&T` and
+    /// `&mut T` may implement `Send` and `Sync`, so the pointer metadata should also
+    /// implement them.
+    ///
+    /// The list of traits should be considered non-exhaustive, as more auto traits may be
+    /// added there in the future. For that reason, it is not allowed to use the metadata
+    /// in a generic context as there is no way to preemptively satisfy those extensions
+    /// to the trait bound.
+    ///
+    /// ```rust,ignore
+    /// The following function does not compile successfully:
+    /// fn generic_metadata<T, U>()
+    /// where
+    ///     T: Pointee<Meta = U> + ?Sized,
+    ///     U: 'static + Copy + Eq + Ord + Send + Sync + Unpin,
+    /// {}
+    /// ```
+    type Meta: 'static + Copy + Eq + Ord + Send + Sync + Unpin + …;
+}
+
+/// Types with a dynamic size.
```

Comment:
It took me forever to understand the reasoning behind the existence of this when it has no items of its own, until I read more closely and noticed that extern types are the only things that are excluded. Would recommend making that clear in the summary for this so that it's more immediately apparent.